playwright-junit/src/main/java/com/microsoft/playwright/BrowserExtension.java

```diff
@@ -0,0 +1,78 @@
+package com.microsoft.playwright;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.microsoft.playwright.ExtensionUtils.hasProperAnnotation;
+import static com.microsoft.playwright.PlaywrightExtension.getOrCreatePlaywright;
+import static org.junit.platform.commons.support.AnnotationSupport.findAnnotation;
+
+public class BrowserExtension implements ParameterResolver {
+  private static final ThreadLocal<Map<Class<? extends BrowserFactory>, Browser>> threadLocalBrowserMap;
+
+  static {
+    threadLocalBrowserMap = ThreadLocal.withInitial(HashMap::new);
+  }
+
+  @Override
+  public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+    if (!hasProperAnnotation(extensionContext)) {
+      return false;
+    }
+    Class<?> clazz = parameterContext.getParameter().getType();
+    return Browser.class.equals(clazz);
+  }
+
+  @Override
+  public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+    BrowserFactory factory = getBrowserFactoryInstance(extensionContext);
+    return getOrCreateBrowser(factory);
+  }
+
+  public static Browser getOrCreateBrowser(BrowserFactory factory) {
+    Class<? extends BrowserFactory> browserFactoryClass = factory.getClass();
+
+    if (getBrowser(browserFactoryClass) != null) {
+      return getBrowser(factory.getClass());
+    }
+
+    Playwright playwright = getOrCreatePlaywright();
+    Browser browser = factory.newBrowser(playwright);
+    System.out.println("Created Browser " + browser);
+    threadLocalBrowserMap.get().put(browserFactoryClass, browser);
+    return browser;
+  }
+
+  static void closeAllBrowsers() {
```

Comment:
Let's close the browser in afterAll callback in this class. Also if you wan to use per method annotation for browser factory, the browser should be closed after such test method finished running (aka afterEach). But I'd keep things simple for now and only support class-level `UseBrowserFactory` annotations, WDYT?

Reply:
I think we don't don't currently want to support method-level parallelizations then we shouldn't support method-level custom factory. That way we don't have to deal with any after each callbacks